### ISSUE TITLE
RDF* support

### ISSUE THREAD

**Issue by rubensworks** [rdfjs/data-model-spec#162]
[RDF*](https://blog.liu.se/olafhartig/2019/01/10/position-statement-rdf-star-and-sparql-star/) provides a way to annotate triples, and is quickly gaining traction by major SPARQL and RDF vendors+frameworks, such as [GraphDB](http://graphdb.ontotext.com/documentation/9.2/free/devhub/rdf-sparql-star.html) and [Jena](https://jena.apache.org/documentation/rdfstar/).

Since RDF/JS aims to be a common data model for RDF applications in JS,
we should consider adding RDF* support to the data model.
Deciding whether or not RDF* is the proper way forward for RDF should not be a concern here IMO.

To enable this, the only thing that would require a change AFAICS, is that the `Quad` interface should implement the `Term` interface (and a `"Quad"` `termType` should be added).

~~Before I open a PR for this, I first want to hear other people's thoughts on this topic.~~

-----

At the moment, two interface proposals seem to be receiving most attention:

* Composition (#163): `QuadTerm` is a `Term` containing a `Quad`.
* Extension (#165): `Quad` implements `Term`.

So far, none of them seem to jump out as a clear "best" approach.
In this comment, I will attempt to summarize the main discussion points (in no particular order), and how they relate to each approach.

|        													  | Composition | Extension |
| ----------------------------------------------------------- | ----------- | --------- |
| Fully backwards compatible with current RDF/JS spec version |             | ✅        |
| Follows RDF* idea that a quad is a term                     |             | ✅        |
| Dynamically check RDF* support on a factory 				  | ✅          |           |
| No optional fields 										  | ✅          |           |
| Alignment to Jena          								  | ✅          |           |
| Directly compatible with non-RDF* factories                 | ✅          |           |

Please let me know if there are any points that are missing from this table, or are missing.

**Comment by RubenVerborgh** [comment]
Sounds good (also the proposed implementation), but we might want it as opt-in.
So not extend `Quad` itself, but rather… `QuadTerm` perhaps?

**Comment by rubensworks** [comment]
> not extend Quad itself, but rather… QuadTerm perhaps?

Yep, that makes sense. Would also be better for backwards-compatibility.
This should align nicely with the TS typings.

**Comment by bergos** [comment]
What about allowing `termType` to be also `undefined` for `Quad`? Would keep the spec small and simple and I don't see any drawback.

**Comment by RubenVerborgh** [comment]
^ That is an interesting suggestion, that seems to match RDF*.

**Comment by rubensworks** [comment]
> What about allowing termType to be also undefined for Quad?

Good idea!

The only drawback I see with this is that it can become trickier to identify whether or not a given object (of unknown type) is a term.
This will for example not work anymore:
```js
function isTerm(object) {
  return !!object.termType;
}
```

So if we go forward with this, we need a different way of identifying terms.

**Comment by bergos** [comment]
Maybe my proposal was not clear enough. I would allow `undefined` and as a string `quad` as `termType` values.

It should not be a problem with that definition. I guess if you want to handle RDF* data, you will use a factory that explicitly supports RDF*. That means there are two cases where `termType` of a `Quad` is `undefined`:

- legacy libraries, not yet updated for RDF*
- new libraries that decided to not support RDF*

For both cases I would say it's right that your `isTerm` function returns false, because it's an explicit choice or it's a missing feature.

**Comment by RubenVerborgh** [comment]
> ```js
>   return !!object.termType;
> ```

BTW, we likely want this to be `typeof object.termType === 'string'`.
Not only is such a check more accurate, it can also be better for performance (because this can be compiled away, or only requires a datatype lookup as opposed to a value lookup and subsequent casting to boolean).

**Comment by rubensworks** [comment]
> I would allow undefined and as a string quad as termType values.

@bergos I misunderstood your earlier comment indeed, my bad.
This makes sense to me for you legacy reasons you pointed out.

Btw, do we want `@rdfjs/data-model` to be a factory that implements RDF* support?
Or shall we fork it to create a dedicate factory, such as `@rdfjs/data-model-star`?

**Comment by retog** [comment]
I don't understand the issue with termType if quadTerm != quad.

A `quadTerm` would always be of termType "quad" and have the quad as its value.

**Comment by rubensworks** [comment]
> I don't understand the issue with termType if quadTerm != quad.

There's no issue anymore AFAICS.

The new range of `termType` would become: "NamedNode", "BlankNode", "Literal", "Variable", "DefaultGraph", "Quad", `undefined`.
The addition of `undefined` here is needed because quads that are produced by old legacy datafactories, will not yet include this `termType` field.

**Comment by bergos** [comment]
@rubensworks I will add RDF* support to `@rdfjs/data-model`, but I will split the exports of the tests to allow running RDF and RDF* test separate.

**Comment by rubensworks** [comment]
@bergos Great!

**Comment by retog** [comment]
> The addition of `undefined` here is needed because quads that are produced by old legacy datafactories, will not yet include this `termType` field.

This seems not be needed if the type `Quad` to be extended to also be a term. Not if there is a separate type `QuadTerm`. I didn't read any objections aganst [that proposal](https://github.com/rdfjs/data-model-spec/issues/162#issuecomment-631291896).

**Comment by Tpt** [comment]
Thank you for this RDF* addition effort!
Sorry for the very late comment.

I just have semantic question: If the `Quad` structure is reused for encoding the nested RDF triple terms, what would be the meaning of this quad:
```js
factory.quad(
  factory.quadTerm(
    factory.namedNode('http://example.com/nestedSubject"),
    factory.namedNode('http://example.com/nestedPredicate"),
    factory.namedNode('http://example.com/nestedObject"),
    factory.namedNode('http://example.com/nestedGraph"),
  ),
  factory.namedNode('http://example.com/predicate"),
  factory.namedNode('http://example.com/object"),
  factory.namedNode('http://example.com/graph")
)
```
If we add this quad to a store, do we have the triple `ex:nestedSubject ex:nestedPredicate ex:nestedSubject` in both the graphs `ex:nestedGraph` and `ex:graph` or just in `ex:Graph`?

To avoid this problem, what about introducing a `TripleTerm` with the following fields instead of introducing a `QuadTerm`:
```
interface TripleTerm {
  attribute string termType; // = "triple"
  attribute Term value; // = ""
  attribute Term subject;
  attribute Term predicate;
  attribute Term object;
};
```

**Comment by rubensworks** [comment]
> I just have semantic question: If the Quad structure is reused for encoding the nested RDF triple terms, what would be the meaning of this quad:
```
factory.quad(
  factory.quadTerm(
    factory.namedNode('http://example.com/nestedSubject"),
    factory.namedNode('http://example.com/nestedPredicate"),
    factory.namedNode('http://example.com/nestedObject"),
    factory.namedNode('http://example.com/nestedGraph"),
  ),
  factory.namedNode('http://example.com/predicate"),
  factory.namedNode('http://example.com/object"),
  factory.namedNode('http://example.com/graph")
)
```

If I understand RDF* correctly, this would be translated into the following:
```
<<ex:nestedSubject ex:nestedPredicate ex:nestedObject ex:nestedGraph>> ex:predicate ex:object ex:graph.
```
This means that `ex:nestedSubject ex:nestedPredicate ex:nestedObject ex:nestedGraph` is in fact not present in the dataset, only an annotation over it is present.
If you also want this quad to be present, the you would have to add it explicitly to the dataset, next to its annotation, like this:
```
<<ex:nestedSubject ex:nestedPredicate ex:nestedObject ex:nestedGraph>> ex:predicate ex:object ex:graph.
ex:nestedSubject ex:nestedPredicate ex:nestedObject ex:nestedGraph.
```

So in summary, nested statements do not inherent named graph containment.

**Comment by Tpt** [comment]
> This means that `ex:nestedSubject ex:nestedPredicate ex:nestedObject ex:nestedGraph` is in fact not present in the dataset, only an annotation over it is present.

Ok! [Foundations of an Alternative Approach to Reification in RDF](https://arxiv.org/pdf/1406.3399.pdf) suggests otherwise by asserting also the nested triples. This seems to be to be a debated point. @hartig [talked about introducing two RDF*" modes in the RDF* mailing list](https://lists.w3.org/Archives/Public/public-rdf-star/2019Sep/0051.html), one with the nested facts also present (just like propery graphs) and where they are not (what you are talking about).

> If I understand RDF* correctly, this would be translated into the following:
> `<<ex:nestedSubject ex:nestedPredicate ex:nestedObject ex:nestedGraph>> ex:predicate ex:object ex:graph.`

I am not sure to understand this translation in using graphs if nested facts are not automatically asserted (what you suggest). It would mean that "the triple `ex:nestedSubject ex:nestedPredicate ex:nestedObject` that is in the graph `ex:nestedGraph` but actually is not because it is not asserted is the subject `s` of the triple `s ex:predicate ex:object` that is in the graph `ex:graph`. Am I correct? If so, why not just write `<<ex:nestedSubject ex:nestedPredicate ex:nestedObject>> ex:predicate ex:object ex:graph` without a nested graph name?

I tried to find existing work about it. [Foundations of an Alternative Approach to Reification in RDF](https://arxiv.org/pdf/1406.3399.pdf) does not present the ability to add a graph name to the nested triples. Jena and GraphDB do not seem to allow it either (at least in the Turtle*/SPARQL* parsers).

**Comment by rubensworks** [comment]
> @hartig talked about introducing two RDF*" modes in the RDF* mailing list, one with the nested facts also present (just like propery graphs) and where they are not (what you are talking about).

Thanks for clearning this up, I wasn't aware of this discussion.
In any case, whatever mode they decide on, I don't think it will have an impact on the RDFJS spec here.

> Foundations of an Alternative Approach to Reification in RDF does not present the ability to add a graph name to the nested triples.

Good point, RDF* is indeed restricted to triples, and doesn't allow annotating quads.
I do wonder if this is something we want to tackle and restrict at the RDFJS spec-level, or if this something parsers and serializers should handle.
I would think the latter, as there may very well be a need for annotating RDF quads as well in the future.
We could for example add a note for this in the spec.

**Comment by hartig** [comment]
Hi all, not sure whether it is exactly related to your discussion here, but I mention it anyways: You are right, my original papers about the RDF*/SPARQL* approach do not consider a notion of named graphs for RDF*. On the other hand, conceiving such a notion is also trivial. Just take the notion of an RDF dataset (as defined by the RDF data model) and, instead of considering it to contain RDF graphs, make it contain RDF* graphs; voila, you have the notion of an RDF* dataset. In a [recent paper](http://olafhartig.de/files/KeskisarkkaEtAl_RSPQLstar_Semantics2019_Preprint.pdf) we have the formal definition of this (see Definition 1 in that paper).

Also, @rubensworks, I am not sure what you mean when you talk about "annotating quads" or "RDF quads." There is no such thing as a "quad" in the RDF data model.

**Comment by rubensworks** [comment]
> You are right, my original papers about the RDF*/SPARQL* approach do not consider a notion of named graphs for RDF*. On the other hand, conceiving such a notion is also trivial.

Great to hear, sounds like supporting that in RDFJS is a good then.

> "annotating ..."

I just imply RDF*, as I feel like it's a bit more descriptive than the name "RDF*" :-)
Happy to use something else if you think "annotation" does not correctly describe RDF*'s intended functionality.

> "RDF quads."

In the [RDFJS spec](http://rdf.js.org/data-model-spec/#quad-interface), we use the notion of a quad as a foundational building block. This roughly corresponds to line in the N-Quads syntax.

**Comment by hartig** [comment]
> In the RDFJS spec, we use the notion of a quad as a foundational building block. This roughly corresponds to line in the N-Quads syntax.

Thanks for the clarification. I was not aware of that.
Now given this, here is an additional response to [your earlier comment](https://github.com/rdfjs/data-model-spec/issues/162#issuecomment-656064505): Conceptually, we may understand your [Quad](http://rdf.js.org/data-model-spec/#quad-interface) data structure as a representation for *an RDF triple within the context of a specific named graph*. In this sense, you are right, while RDF* can be used to make statements about individual triples (independent of where these triples may occur), but it cannot be used to make statements about a specific occurrence of such a triple.

**Comment by Tpt** [comment]
Thank you @hartig and @rubensworks for your replies!

> Now given this, here is an additional response to your earlier comment: Conceptually, we may understand your Quad data structure as a representation for an RDF triple within the context of a specific named graph. In this sense, you are right, while RDF* can be used to make statements about individual triples (independent of where these triples may occur), but it cannot be used to make statements about a specific occurrence of such a triple.

Thank you! So, if understand correctly,
```
<<ex:nestedSubject ex:nestedPredicate ex:nestedObject>> ex:predicate ex:object ex:graph.
```
has a meaning and is valid in NQuads*. And
```
<<ex:nestedSubject ex:nestedPredicate ex:nestedObject ex:nestedGraph>> ex:predicate ex:object ex:graph .
```
is invalid (option 1), or could be considered to be the same as
```
<<ex:nestedSubject ex:nestedPredicate ex:nestedObject>> ex:predicate ex:object ex:graph .
```
in "single assertion mode" or to be the same as
```
<<ex:nestedSubject ex:nestedPredicate ex:nestedObject>> ex:predicate ex:object ex:graph .
ex:nestedSubject ex:nestedPredicate ex:nestedObject ex:nestedGraph .
```
in "property graph mode" (option 2).
Am I correct?

If yes. I would be in favor of adding to the RDF/JS model specification a note stating that the implementations should follow the option (2) behaviors. If it is fine for everyone, I could maybe do a pull request with it as soon as one of the existing "quad term" proposal is adopted.

**Comment by hartig** [comment]
@Tpt Note that we do not have a definition of a serialization format called NQuads*. In fact, we don't even have an RDF*-specific version of the [N-Triples](https://www.w3.org/TR/n-triples/) format. However, the latter may be extended easily into something that we may call N-Triples* along the lines of how we have extended [Turtle](https://www.w3.org/TR/turtle/) into Turtle*. Then, it is not difficult to see how this N-Triples* format may be extended further into a format called N-Quads*. In this format, you may then write something like your first example line:
> <<ex:nestedSubject ex:nestedPredicate ex:nestedObject>> ex:predicate ex:object ex:graph.

On the other hand, your second example line
> <<ex:nestedSubject ex:nestedPredicate ex:nestedObject ex:nestedGraph>> ex:predicate ex:object ex:graph .

would not make any sense at all in this context and, thus, should not be a valid expression in that hypothetical N-Quads* syntax. The reason is that there is no way how this can be parsed into an RDF* graph (or into an RDF* dataset if you want); the RDF* data model simply does not contain a notion of nested 4-tuples (exactly like the RDF data model which also does not contain any notion of 4-tuples).

**Comment by Tpt** [comment]
@hartig Thank you for your comment and sorry for not being clear and for the notation abuses.

I abused the NQuad format to represent a problem that the currently proposed RDF/JS APIs for RDF* could lead to (c.f. #163 #164 #165) without having to write JavaScript code. The 3 proposals all allow a variation of the following code that leads to the case presented by my second example line:
```javascript
factory.quad(
  factory.quadTerm(
    factory.namedNode('http://example.com/nestedSubject"),
    factory.namedNode('http://example.com/nestedPredicate"),
    factory.namedNode('http://example.com/nestedObject"),
    factory.namedNode('http://example.com/nestedGraph"),
  ),
  factory.namedNode('http://example.com/predicate"),
  factory.namedNode('http://example.com/object"),
  factory.namedNode('http://example.com/graph")
)
```
(note the graph name in the nested "quad"). It is why I tried to find a proper way to deal with this case.

My option 1 is to forbid this completely by making the `quadTerm` constructor fail if a graph name is provided. This follows what you said. However, this requires the API user to think about clearing the graph name if they they reuse a quad from e.g. a SPARQL query. My option 2 is to ignore the "graph name" of the nested quad except if the inner triples should also be automatically asserted ("SA mode"). This way, there is no ambiguity and no need for the user to check that the quad graph name field is empty. What do you think about it?

**Comment by hartig** [comment]
Thomas, thanks a lot for this clarification (I am clearly not familiar with RDF/JS and the terminology that you use in this community ;-)

I understand see the dilemma now (which, IMHO, originates from the desire to build everything around this misleading [`Quad`](http://rdf.js.org/data-model-spec/#quad-interface) data structure which is conceptually incorrect even from a pure RDF perspective).

Option 1 may still be the cleanest option because it makes users aware of what they are doing. By the way, I don't see how such a `Quad` object with a non-null `graph` attribute could ever result from a SPARQL query. SELECT queries return sets of solution mappings and CONSTRUCT queries return triples. However, I see that there may be other ways in which an API user may end up with such a `Quad` object and then try to (incorrectly) make an RDF* triple about it.

Option 2 seems tricky to me. Simply ignoring whether the `graph` attribute is null or not is misleading and, in fact, incorrect from an RDF* perspective. So, it my lead to a lot of confusion and to weird behavior in applications.

Another option might be to silently set the `graph` attribute to null, but that does not seem to be desirable either.

**Comment by gkellogg** [comment]
This does raise RDF* data model issues though. Ruby RDF also uses quads internally, which I think is a fairly common problem. My interpretation is that RDF* is a triple annotation format, and not a quad format. But, if a statement inside a named graph has a triple as a subject or object, and were presuming PG mode, where is that triple to be assured. My view has been that the triple would be asserted in the default graph, but I’m increasingly becoming wary of PG mode, and find the separate assertions mode conceptually cleaner. That way, if a triple is intended to be asserted in a named graph, it can just be stated as such by the author. 

It gets more interesting in the JSON-LD* work we’ve started. PG mod seriously complicates algorithms such as expansion and compaction.

**Comment by hartig** [comment]
Gregg, that's a very interesting observation and one that I have never thought about (primarily because I have focused mostly on the single graph case so far). I agree that this is a strong point against PG mode!

**Comment by rubensworks** [comment]
At the moment, two interface proposals seem to be receiving most attention:

* Composition (#163): `QuadTerm` is a `Term` containing a `Quad`.
* Extension (#165): `Quad` implements `Term`.

So far, none of them seem to jump out as a clear "best" approach.
In the original issue [above](https://github.com/rdfjs/data-model-spec/issues/162), I have summarized the main discussion points (in no particular order), and how they relate to each approach.
Please let me know if there are any points that are missing from this table, or are missing.

**Comment by rubensworks** [comment]
Can be closed, since #165 has been merged.
